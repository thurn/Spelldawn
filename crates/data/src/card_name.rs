// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use convert_case::{Case, Casing};
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use strum_macros::Display;

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Display, Serialize, Deserialize)]
pub enum CardName {
    /// Empty identity cards with no associated rules text, for use in tests
    TestChampionIdentity,
    TestOverlordIdentity,

    ArcaneRecovery,
    Greataxe,
    GoldMine,
    IceDragon,
    DungeonAnnex,
}

impl CardName {
    pub fn displayed_name(&self) -> String {
        format!("{}", self).from_case(Case::Pascal).to_case(Case::Title)
    }
}

impl PartialOrd<Self> for CardName {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for CardName {
    fn cmp(&self, other: &Self) -> Ordering {
        self.to_string().cmp(&other.to_string())
    }
}
