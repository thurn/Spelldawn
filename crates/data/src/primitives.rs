// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Fundamental types and data structures for Spelldawn

use std::fmt;

use serde::{Deserialize, Serialize};

pub type TurnNumber = u32;
pub type ActionCount = u32;
pub type ManaValue = u32;
pub type PointsValue = u32;
pub type HealthValue = u32;
pub type AttackValue = u32;
pub type ShieldValue = u32;
pub type BoostCount = u32;
pub type LevelValue = u32;

/// Identifies a player across different games
#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
pub struct PlayerId {
    pub value: u64,
}

impl PlayerId {
    pub fn new(value: u64) -> Self {
        Self { value }
    }

    /// Byte array representation of this ID
    pub fn key(&self) -> [u8; 8] {
        self.value.to_be_bytes()
    }
}

impl fmt::Debug for PlayerId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Identifies an ongoing game
#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
pub struct GameId {
    pub value: u64,
}

impl GameId {
    pub fn new(value: u64) -> Self {
        Self { value }
    }

    /// Byte array representation of this ID
    pub fn key(&self) -> [u8; 8] {
        self.value.to_be_bytes()
    }
}

impl fmt::Debug for GameId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// The two players in a game: Overlord & Champion
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub enum Side {
    Overlord,
    Champion,
}

impl Side {
    /// Gets the opponent of the provided player
    pub fn opponent(&self) -> Self {
        match self {
            Side::Champion => Self::Overlord,
            Side::Overlord => Self::Champion,
        }
    }
}

impl fmt::Debug for Side {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Side::Overlord => "Overlord",
                Side::Champion => "Champion",
            }
        )
    }
}

/// Identifies a card in an ongoing game
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub struct CardId {
    pub side: Side,
    pub index: usize,
}

impl CardId {
    pub fn new(side: Side, index: usize) -> Self {
        Self { side, index }
    }
}

impl fmt::Debug for CardId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}",
            match self.side {
                Side::Overlord => "O",
                Side::Champion => "C",
            },
            self.index
        )
    }
}

/// Identifies an ability position within a card's 'abilities' vector
#[derive(PartialEq, Eq, Hash, Copy, Clone, Ord, PartialOrd, Serialize, Deserialize)]
pub struct AbilityIndex(pub usize);

impl fmt::Debug for AbilityIndex {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// Identifies an ability within a card. Abilities are the only game entity
/// which may contain delegates..
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub struct AbilityId {
    pub card_id: CardId,
    pub index: AbilityIndex,
}

impl fmt::Debug for AbilityId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}[{:?}]", self.card_id, self.index)
    }
}

impl AbilityId {
    pub fn new(card_id: CardId, index: usize) -> Self {
        Self { card_id, index: AbilityIndex(index) }
    }
}

/// Uniquely identifies a raid within a given game
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub struct RaidId(pub u32);

impl fmt::Debug for RaidId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "%{}", self.0)
    }
}

/// Contains the URL of an image asset within a game
#[derive(PartialEq, Eq, Hash, Debug, Clone, Serialize, Deserialize)]
pub struct Sprite {
    pub address: String,
}

impl Sprite {
    pub fn new(address: impl Into<String>) -> Self {
        Self { address: address.into() }
    }
}

/// The schools of magic, which provide restrictions on players during
/// deckbuilding
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum School {
    Neutral,
    Shadow,
    Nature,
    Time,
}

/// The possible Rooms in which the Overlord player may play their cards.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum RoomId {
    /// The Overlord's deck
    Vault,
    /// The Overlord's hand
    Sanctum,
    /// The Overlord's discard pile
    Crypts,
    RoomA,
    RoomB,
    RoomC,
    RoomD,
    RoomE,
}

impl RoomId {
    /// An 'inner room' is one of the three predefined rooms for the Overlord
    /// player's deck, hand, and discard pile. Inner rooms cannot contain
    /// Schemes or Projects.
    pub fn is_inner_room(&self) -> bool {
        matches!(self, RoomId::Vault | RoomId::Sanctum | RoomId::Crypts)
    }
}

/// Used to control where a card is rendered within a room
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum RoomLocation {
    Defender,
    Occupant,
}

/// Used to control where an item is rendered within the Champion's item display
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum ItemLocation {
    Weapons,
    Artifacts,
}

/// The three possible factions of weapons and minions. Minions can only be
/// damaged by weapons from the same faction.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum Faction {
    Mortal,
    Abyssal,
    Infernal,
}

/// Rarity of a card, used to determine how likely it is to appear in randomized
/// rewards.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum Rarity {
    Common,
    Uncommon,
    Rare,
    Epic,

    /// Card cannot be obtained via random rewards
    None,
}

/// Possible types of cards
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum CardType {
    Spell,
    Weapon,
    Artifact,
    Minion,
    Project,
    Scheme,
    Upgrade,
    Identity,
}

/// Subtypes of cards
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum CardSubtype {
    Silvered,
}

/// Describes a boost ability activation
#[derive(PartialEq, Eq, Hash, Debug, Clone, Copy, Serialize, Deserialize)]
pub struct BoostData {
    /// Boosted card
    pub card_id: CardId,
    /// How many times was the boost applied?
    pub count: u32,
}
