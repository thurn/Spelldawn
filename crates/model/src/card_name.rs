// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use strum_macros::Display;

use convert_case::{Case, Casing};

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Display)]
pub enum CardName {
    /// Empty identity cards with no associated rules text, for use in unit tests
    TestChampionIdentity,
    TestOverlordIdentity,

    ArcaneRecovery,
    Greataxe,
    GoldMine,
    IceDragon,
    DungeonAnnex,
}

impl CardName {
    pub fn displayed_name(&self) -> String {
        format!("{}", self).from_case(Case::Pascal).to_case(Case::Title)
    }
}
