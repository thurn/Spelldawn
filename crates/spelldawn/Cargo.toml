[package]
name = "spelldawn"
version = "0.0.0"
authors = ["Spelldawn Team"]
homepage = "https://github.com/thurn/spelldawn"
description = "An asymmetric trading card game"
license = "Apache-2.0"
default-run = "spelldawn"
edition = "2021"
rust-version = "1.57"

[[bin]]
name = "spelldawn"
path = "src/main.rs"

[dependencies]
prost = "0.9.0"
prost-types = "0.9.0"
tokio = { version = "1.14.0", features = ["macros", "rt-multi-thread", "sync", "time"] }
tokio-stream = "0.1.8"
tonic = "0.6.1"
tonic-web = "0.2"

protos = { path = "../protos", version = "0.0.0" }