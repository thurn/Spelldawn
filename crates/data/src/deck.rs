// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Defines a deck as it exists outside of an active game

use std::collections::HashMap;
use std::iter;

use crate::card_name::CardName;
use crate::primitives::UserId;

/// Represents a player deck outside of an active game
#[derive(Debug, Clone)]
pub struct Deck {
    /// Identifies the user who owns this deck
    pub owner_id: UserId,
    /// Identity card for this deck
    pub identity: CardName,
    /// How many (non-identity) cards with each name are present in this deck?
    pub cards: HashMap<CardName, u32>,
}

impl Deck {
    /// Returns a vector which repeats each [CardName] in [Self::cards] in
    /// alphabetical order a number of times equal to its deck count. Note: The
    /// returned vector does *not* contain [Self::identity].
    pub fn card_names(&self) -> Vec<CardName> {
        let mut result = self
            .cards
            .iter()
            .flat_map(|(name, count)| iter::repeat(*name).take(*count as usize))
            .collect::<Vec<_>>();
        result.sort();
        result
    }
}
