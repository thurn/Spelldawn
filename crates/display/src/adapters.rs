// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Helpers for converting between server & client representations

use anyhow::{anyhow, bail, Context, Result};
use data::primitives::{AbilityId, AbilityIndex, CardId, GameId, PlayerId, RoomId, Side};
use protos::spelldawn::game_object_identifier::Id;
use protos::spelldawn::{
    CardIdentifier, GameIdentifier, GameObjectIdentifier, PlayerIdentifier, PlayerName, PlayerSide,
    RoomIdentifier,
};

pub fn adapt_game_id(game_id: GameId) -> GameIdentifier {
    GameIdentifier { value: game_id.value }
}

/// Converts a [Side] into a [PlayerName] based on which viewer we are rendering
/// this update for.
pub fn to_player_name(side: Side, user_side: Side) -> PlayerName {
    if side == user_side {
        PlayerName::User
    } else {
        PlayerName::Opponent
    }
}

pub fn adapt_player_id(player_id: PlayerId) -> PlayerIdentifier {
    PlayerIdentifier { value: player_id.value }
}

pub fn adapt_side(side: Side) -> PlayerSide {
    match side {
        Side::Overlord => PlayerSide::Overlord,
        Side::Champion => PlayerSide::Champion,
    }
}

pub fn to_server_side(side: Option<PlayerSide>) -> Result<Side> {
    match side {
        Some(PlayerSide::Overlord) => Ok(Side::Overlord),
        Some(PlayerSide::Champion) => Ok(Side::Champion),
        _ => bail!("Invalid side"),
    }
}

/// Turns a server [CardId] into its protobuf equivalent
pub fn adapt_card_id(card_id: CardId) -> CardIdentifier {
    // TODO: Obfuscate this somehow, directly using the index leaks information
    CardIdentifier {
        side: match card_id.side {
            Side::Overlord => PlayerSide::Overlord,
            Side::Champion => PlayerSide::Champion,
        }
        .into(),
        index: card_id.index as u32,
        ability_id: None,
    }
}

/// Turns a server [CardId] and [AbilityIndex] into a protobuf equivalent
pub fn adapt_ability_card_id(card_id: CardId, ability_index: AbilityIndex) -> CardIdentifier {
    CardIdentifier { ability_id: Some(ability_index.value() as u32), ..adapt_card_id(card_id) }
}

/// Turns an [AbilityId] into a [CardIdentifier].
pub fn adapt_ability_id(ability_id: AbilityId) -> CardIdentifier {
    adapt_ability_card_id(ability_id.card_id, ability_id.index)
}

/// Turns a server [RoomId] into its protobuf equivalent
pub fn adapt_room_id(room_id: RoomId) -> RoomIdentifier {
    match room_id {
        RoomId::Vault => RoomIdentifier::Vault,
        RoomId::Sanctum => RoomIdentifier::Sanctum,
        RoomId::Crypts => RoomIdentifier::Crypts,
        RoomId::RoomA => RoomIdentifier::RoomA,
        RoomId::RoomB => RoomIdentifier::RoomB,
        RoomId::RoomC => RoomIdentifier::RoomC,
        RoomId::RoomD => RoomIdentifier::RoomD,
        RoomId::RoomE => RoomIdentifier::RoomE,
    }
}

pub fn to_server_room_id(identifier: i32) -> Result<RoomId> {
    match RoomIdentifier::from_i32(identifier) {
        Some(RoomIdentifier::Vault) => Ok(RoomId::Vault),
        Some(RoomIdentifier::Sanctum) => Ok(RoomId::Sanctum),
        Some(RoomIdentifier::Crypts) => Ok(RoomId::Crypts),
        Some(RoomIdentifier::RoomA) => Ok(RoomId::RoomA),
        Some(RoomIdentifier::RoomB) => Ok(RoomId::RoomB),
        Some(RoomIdentifier::RoomC) => Ok(RoomId::RoomC),
        Some(RoomIdentifier::RoomD) => Ok(RoomId::RoomD),
        Some(RoomIdentifier::RoomE) => Ok(RoomId::RoomE),
        _ => bail!("Invalid RoomId: {:?}", identifier),
    }
}

#[derive(Copy, Clone, Eq, PartialEq)]
pub enum ServerCardId {
    CardId(CardId),
    AbilityId(AbilityId),
}

impl ServerCardId {
    // Helper for when it is an error for the value to not be a
    // `ServerCardId::CardId`.
    pub fn as_card_id(self) -> Result<CardId> {
        match self {
            ServerCardId::CardId(card_id) => Ok(card_id),
            ServerCardId::AbilityId(_) => bail!("Expected CardId"),
        }
    }
}

/// Converts a client [CardIdentifier] into a server [CardId] or [AbilityId].
pub fn to_server_card_id(card_id: Option<CardIdentifier>) -> Result<ServerCardId> {
    if let Some(id) = card_id {
        let result = CardId {
            side: match id.side() {
                PlayerSide::Overlord => Side::Overlord,
                PlayerSide::Champion => Side::Champion,
                _ => bail!("Invalid CardId {:?}", card_id),
            },
            index: id.index as usize,
        };

        id.ability_id.map_or(Ok(ServerCardId::CardId(result)), |index| {
            Ok(ServerCardId::AbilityId(AbilityId {
                card_id: result,
                index: AbilityIndex(index as usize),
            }))
        })
    } else {
        Err(anyhow!("Missing Required CardId"))
    }
}

/// Converts a [CardId] into a client [GameObjectIdentifier]
pub fn card_id_to_object_id(id: CardId) -> GameObjectIdentifier {
    GameObjectIdentifier { id: Some(Id::CardId(adapt_card_id(id))) }
}

/// Converts a client [PlayerIdentifier] into a server [PlayerId].
pub fn to_server_player_id(player_id: Option<PlayerIdentifier>) -> Result<PlayerId> {
    Ok(PlayerId::new(player_id.with_context(|| "PlayerId is required")?.value))
}
