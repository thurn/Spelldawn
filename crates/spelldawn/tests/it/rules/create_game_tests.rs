// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use data::card_name::CardName;
use data::deck::Deck;
use data::primitives::{GameId, PlayerId};
use display::adapters;
use insta::assert_snapshot;
use maplit::hashmap;
use protos::spelldawn::game_action::Action;
use protos::spelldawn::{CreateNewGameAction, PlayerName, PlayerSide};
use test_utils::client::{HasText, TestSession};
use test_utils::fake_database::FakeDatabase;
use test_utils::summarize::Summary;
use test_utils::*;

#[test]
fn create_new_game() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_test_session(game_id, overlord_id, champion_id);
    let response = session.perform_action_with_game_id(
        Action::CreateNewGame(CreateNewGameAction {
            side: PlayerSide::Overlord.into(),
            opponent_id: Some(adapters::adapt_player_id(session.opponent_id())),
            deterministic: true,
        }),
        session.user_id(),
        None,
    );

    assert_snapshot!(Summary::run(&response));
}

#[test]
fn connect_to_new_game() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_test_session(game_id, overlord_id, champion_id);
    session
        .perform_action_with_game_id(
            Action::CreateNewGame(CreateNewGameAction {
                side: PlayerSide::Overlord.into(),
                opponent_id: Some(adapters::adapt_player_id(session.opponent_id())),
                deterministic: true,
            }),
            session.user_id(),
            None,
        )
        .expect("create game");
    let response = session.connect(overlord_id, Some(game_id));

    assert!(session.user.interface.controls().has_text("Keep"));
    assert!(session.user.interface.controls().has_text("Mulligan"));
    assert_eq!(5, session.user.cards.browser().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_snapshot!(Summary::run(&response));
}

#[test]
fn keep_opening_hand() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_test_session(game_id, overlord_id, champion_id);
    session
        .perform_action_with_game_id(
            Action::CreateNewGame(CreateNewGameAction {
                side: PlayerSide::Overlord.into(),
                opponent_id: Some(adapters::adapt_player_id(session.opponent_id())),
                deterministic: true,
            }),
            session.user_id(),
            None,
        )
        .expect("create game");
    session.connect(overlord_id, Some(game_id)).expect("connect");
    session.connect(champion_id, Some(game_id)).expect("connect");

    let response = session.click_on(overlord_id, "Keep");
    assert!(session.user.interface.controls().has_text("Waiting"));
    assert_eq!(0, session.user.cards.browser().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_eq!(0, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.browser().len());

    assert_snapshot!(Summary::summarize(&response));
}

#[test]
fn mulligan_opening_hand() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_test_session(game_id, overlord_id, champion_id);
    session
        .perform_action_with_game_id(
            Action::CreateNewGame(CreateNewGameAction {
                side: PlayerSide::Overlord.into(),
                opponent_id: Some(adapters::adapt_player_id(session.opponent_id())),
                deterministic: true,
            }),
            session.user_id(),
            None,
        )
        .expect("create game");
    session.connect(overlord_id, Some(game_id)).expect("connect");
    session.connect(champion_id, Some(game_id)).expect("connect");

    let response = session.click_on(overlord_id, "Mulligan");
    assert_snapshot!(Summary::summarize(&response));

    assert!(session.user.interface.controls().has_text("Waiting"));
    assert_eq!(0, session.user.cards.browser().len());
    assert_eq!(5, session.user.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.user.cards.hand(PlayerName::Opponent).len());

    assert_eq!(0, session.opponent.cards.hand(PlayerName::User).len());
    assert_eq!(5, session.opponent.cards.hand(PlayerName::Opponent).len());
    assert_eq!(5, session.opponent.cards.browser().len());
}

#[test]
fn both_keep_opening_hands() {
    let (game_id, overlord_id, champion_id) = generate_ids();
    let mut session = make_test_session(game_id, overlord_id, champion_id);
    session
        .perform_action_with_game_id(
            Action::CreateNewGame(CreateNewGameAction {
                side: PlayerSide::Overlord.into(),
                opponent_id: Some(adapters::adapt_player_id(session.opponent_id())),
                deterministic: true,
            }),
            session.user_id(),
            None,
        )
        .expect("create game");
    session.connect(overlord_id, Some(game_id)).expect("connect");
    session.connect(champion_id, Some(game_id)).expect("connect");

    session.click_on(overlord_id, "Keep");
    let response = session.click_on(champion_id, "Keep");
    assert_snapshot!(Summary::summarize(&response));

    assert_eq!(5, session.user.this_player.mana());
    assert_eq!(5, session.user.other_player.mana());
    assert_eq!(5, session.opponent.this_player.mana());
    assert_eq!(5, session.opponent.other_player.mana());

    assert_eq!(3, session.user.this_player.actions());
    assert_eq!(0, session.user.other_player.actions());
    assert_eq!(0, session.opponent.this_player.actions());
    assert_eq!(3, session.opponent.other_player.actions());

    assert!(session.dusk());
}

fn make_test_session(game_id: GameId, overlord_id: PlayerId, champion_id: PlayerId) -> TestSession {
    let database = FakeDatabase {
        generated_game_id: Some(game_id),
        game: None,
        overlord_deck: Some(Deck {
            owner_id: overlord_id,
            identity: CardName::TestOverlordIdentity,
            cards: hashmap! {CardName::TestOverlordSpell => 45},
        }),
        champion_deck: Some(Deck {
            owner_id: champion_id,
            identity: CardName::TestChampionIdentity,
            cards: hashmap! {CardName::TestChampionSpell => 45},
        }),
    };

    TestSession::new(database, overlord_id, champion_id)
}
