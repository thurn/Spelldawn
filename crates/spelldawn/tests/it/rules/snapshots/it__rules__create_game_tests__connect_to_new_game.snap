---
source: crates/spelldawn/tests/it/rules/create_game_tests.rs
assertion_line: 67
expression: "Summary::run(&response)"

---

    UpdateGameView: 
        user: 
            side: Overlord
            player_info: 
                name: "Test Overlord Identity"
                portrait: "<SpriteAddress>"
                portrait_frame: "<SpriteAddress>"
                card_back: "<SpriteAddress>"
            mana: 0
            action_tracker: 0
            score: 0
        opponent: 
            side: Champion
            player_info: 
                name: "Test Champion Identity"
                portrait: "<SpriteAddress>"
                portrait_frame: "<SpriteAddress>"
                card_back: "<SpriteAddress>"
            mana: 0
            action_tracker: 0
            score: 0
        current_priority: Unspecified
        raid_active: false
    CreateOrUpdateCard: 
        card: 
            card_id: O0
            revealed_to_viewer: true
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Identity"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 0
            position: ObjectPositionIdentity { owner: User }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: O1
            revealed_to_viewer: true
            revealed_to_opponent: false
            card_icons: 
                top_left_icon: "1"
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Spell"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 6
            position: ObjectPositionHand { owner: User }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: O2
            revealed_to_viewer: true
            revealed_to_opponent: false
            card_icons: 
                top_left_icon: "1"
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Spell"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 7
            position: ObjectPositionHand { owner: User }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: O3
            revealed_to_viewer: true
            revealed_to_opponent: false
            card_icons: 
                top_left_icon: "1"
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Spell"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 8
            position: ObjectPositionHand { owner: User }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: O4
            revealed_to_viewer: true
            revealed_to_opponent: false
            card_icons: 
                top_left_icon: "1"
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Spell"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 9
            position: ObjectPositionHand { owner: User }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: O5
            revealed_to_viewer: true
            revealed_to_opponent: false
            card_icons: 
                top_left_icon: "1"
            arena_frame: "<SpriteAddress>"
            owning_player: User
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Overlord Spell"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 10
            position: ObjectPositionHand { owner: User }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: C0
            revealed_to_viewer: true
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Champion Identity"
                rules_text: "<RulesText>"
                targeting: 
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionStaging
                can_play: false
        create_position: 
            sorting_key: 0
            position: ObjectPositionIdentity { owner: Opponent }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: C1
            revealed_to_viewer: false
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
        create_position: 
            sorting_key: 16
            position: ObjectPositionHand { owner: Opponent }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: C2
            revealed_to_viewer: false
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
        create_position: 
            sorting_key: 17
            position: ObjectPositionHand { owner: Opponent }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: C3
            revealed_to_viewer: false
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
        create_position: 
            sorting_key: 18
            position: ObjectPositionHand { owner: Opponent }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: C4
            revealed_to_viewer: false
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
        create_position: 
            sorting_key: 19
            position: ObjectPositionHand { owner: Opponent }
        create_animation: Unspecified
    CreateOrUpdateCard: 
        card: 
            card_id: C5
            revealed_to_viewer: false
            revealed_to_opponent: true
            card_icons: 
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
        create_position: 
            sorting_key: 20
            position: ObjectPositionHand { owner: Opponent }
        create_animation: Unspecified
    RunInParallel: 
        MoveGameObjects: 
            id: O1
            position: 
                sorting_key: 6
                position: ObjectPositionBrowser
        MoveGameObjects: 
            id: O2
            position: 
                sorting_key: 7
                position: ObjectPositionBrowser
        MoveGameObjects: 
            id: O3
            position: 
                sorting_key: 8
                position: ObjectPositionBrowser
        MoveGameObjects: 
            id: O4
            position: 
                sorting_key: 9
                position: ObjectPositionBrowser
        MoveGameObjects: 
            id: O5
            position: 
                sorting_key: 10
                position: ObjectPositionBrowser
    RenderInterface: 
    panels: 
        main_controls: 
            node: 
                text: "Keep"
                text: "Mulligan"
        card_anchor_nodes: 
    RenderInterface: 
    panels: "<DebugPanel>"
